system/services/websiteUsers/WebsiteUserImpersonationService.cfc

```diff
@@ -0,0 +1,74 @@
+/**
+ * @singleton
+ * @presideservice
+ * @autodoc
+ */
+component displayName="Website user impersonation service" {
+
+// constructor
+	/**
+	 * @websiteLoginService.inject   websiteLoginService
+	 */
+	public any function init( required any websiteloginService ) {
+		_setWebsiteLoginService( arguments.websiteLoginService );
+		variables.impersonations = {};
+
+		return this;
+	}
+
+// public api methods
+	public string function create( required string userId, required string targetUrl ) {
+		var impersonationId  = _addImpersonation( argumentCollection=arguments );
+		var impersonationUrl = arguments.targetUrl;
+
+		if ( find( "?", impersonationUrl ) ) {
+			impersonationUrl &= "&impersonate=" & impersonationId;
+		} else {
+			impersonationUrl &= "?impersonate=" & impersonationId;
+		}
+
+		return impersonationUrl;
+	}
+
+	public string function resolve( required string id ) {
+		var impersonation = _getImpersonation( arguments.id );
+
+		if ( impersonation.isEmpty() ) {
+			return "";
+		}
+
+		_getWebsiteLoginService().impersonate( impersonation.userId );
+		_removeImpersonation( arguments.id );
+
+		return impersonation.targetUrl;
+	}
+
+
+// private methods
+	private string function _addImpersonation( required string userId, required string targetUrl ) {
+		var impersonationId = createUUID();
+
+		variables.impersonations[ impersonationId ] = {
```

Comment:
The only issue with using variables scope for this type of caching is that this feature will not be clusterable. Some variables caching of features are ok, because they are just warming up computation caching that will be consistent across all nodes - but in this case, you would not be able to have Preside in a cluster and configure this impersonation to work consistently. 

Suggest refactoring so that the service accepts a cachebox cache and uses that to store the impersonations. (if using a cluster, can then configure the cache to use an eternal cache provider).

Reply:
Done and committed...